### PR TITLE
Allows a scene to trigger another scene

### LINKED ISSUE #797: Scene: be able to start a scene from a scene
⚠️  The user shouldn't be able to select the same scene !!

### PR DESCRIPTION
**Pull request by rob-mccann**
### Pull Request check-list

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] If your changes modify the API (REST or Node.js), did you modify the API documentation? (Documentation is based on comments in code)
- [x] If you are adding a new features/services which needs explanation, did you modify the user documentation? See [the GitHub repo](https://github.com/GladysAssistant/v4-website) and the [website](https://gladysassistant.com).
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.json`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

### Description of change

This change adds an action to scenes which triggers another scene.

image

A few assumptions have been made here which I'd be happy to revisit based on feedback:

- There's no validation that a scene exists in the backend
- A scene can only trigger another scene once across all action groups
  - In the cyclical scenario (for example,  Scene A that triggers Scene B that triggers Scene A), each scene will be called once only and a warning message logged. This is handled at execution time by remembers which scenes were already called.
  - There's no cyclical checks/validation done when a scene is created at the moment; I suggest this is handled in a separate PR to make this easier to review and in the interest of making it incrementally better :) 
- At the moment, there isnt trigger shown in the scene being called

If you're happy with the approach, I'll go ahead and update the docs.

Closes #797

### CODE REVIEW

File: server/lib/scene/scene.execute.js [hunk 1]
```diff
@@ -8,11 +8,15 @@ const logger = require('../../utils/logger');
...
  * sceneManager.execute('test');
  */
-function execute(sceneSelector, scope) {
+function execute(sceneSelector, scope = {}) {
   try {
     if (!this.scenes[sceneSelector]) {
       throw new Error(`Scene with selector ${sceneSelector} not found.`);
     }
+
+    scope.sceneCircularDependencyPreventionList = scope.sceneCircularDependencyPreventionList || new Set();
```
Comment (atrovato) [inline server/lib/scene/scene.execute.js]: Hum... I'm not really fan of `sceneCircularDependencyPreventionList` name, a bit long, and is a Set, not a list.
Maybe just `executedScenes`, or `relativeScenes`?

File: server/lib/scene/scene.actions.js [hunk 2]
```diff
@@ -103,7 +103,16 @@ const actionsFunc = {
       }
       setTimeout(resolve, timeToWaitMilliseconds);
     }),
-  [ACTIONS.SCENE.START]: async (self, action, scope) => self.execute(action.scene, scope),
+  [ACTIONS.SCENE.START]: async (self, action, scope) => {
+    if (scope.sceneCircularDependencyPreventionList.has(action.scene)) {
```
Comment (atrovato) [inline server/lib/scene/scene.actions.js]: That's nice to check circular execution, but how can we alert user in this case?
Maybe we should add a "warning" message in the scene editor to explain that each depend scenes will be run only once.

**Review by atrovato — COMMENTED** [review]
Tested locally, looks fine.
2 things:
 - we may want to add a message in the UI to tell the user that scenes won't loop
 - execution conditions of the child scene are ignored (to execute it), maybe the user needs to know it too

what about a ℹ️ tooltip on scene selection box?

File: server/lib/scene/scene.execute.js (hunk 1 above)
Comment (rob-mccann) [reply server/lib/scene/scene.execute.js]: Naming things is hard 😓 

`scenesAlreadyExecuted` perhaps

File: server/lib/scene/scene.actions.js (hunk 2 above)
Comment (Pierre-Gilles) [reply server/lib/scene/scene.actions.js]: I think the box in the UI could have a little description explaining that circular execution is prevented ? Just before the select

File: server/lib/scene/scene.execute.js (hunk 1 above)
Comment (atrovato) [reply server/lib/scene/scene.execute.js]: I know... naming... `alreadyExecutedScenes` instead.

Comment (rob-mccann) [reply server/lib/scene/scene.execute.js]: done

File: server/lib/scene/scene.actions.js (hunk 2 above)
Comment (rob-mccann) [reply server/lib/scene/scene.actions.js]: I've added a message - any suggestions for copy and/or 🇫🇷 ?

Comment (atrovato) [reply server/lib/scene/scene.actions.js]: According to your message `"Scenes will only be called once in a chain, even if they're added multiple times. Trigger conditions will be ignored."`
"Chaque scène ne sera exécutée qu'une seule fois, et les conditions d'exécution seront ignorées."

@Pierre-Gilles ?

Comment (Pierre-Gilles) [reply server/lib/scene/scene.actions.js]: I would say in french: 

"Les conditions de déclenchements de la scène appelée seront ignorées. Si une scène appelle une autre scène, et vice-versa, Gladys ne tombera pas dans une boucle infinie: chaque scène ne sera exécutée qu'une seule fois. "

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]
When I click on the start scene button: 

Screenshot 2021-04-22 at 12 20 09

I got this error: 

Screenshot 2021-04-22 at 12 20 30

I guess it's because `alreadyExecutedScenes` does not exist at this stage

**Review by Pierre-Gilles — APPROVED** [review]
Looks good to me this time! Thanks for the PR, great work 🙏

**Pull request merged**